File: lintcode/search-a-2d-matrix-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/search-a-2d-matrix-ii/

class Solution:
    """
    @param matrix: An list of lists of integers
    @param target: An integer you want to search in matrix
    @return: An integer indicates the total occurrence of target in the given matrix
    """
    def searchMatrix(self, matrix, target):
        # write your code here
        ans = 0
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return ans

        x = 0
        y = len(matrix[0]) - 1
        while x < len(matrix):
            while y >= 0 and matrix[x][y] > target:
                y -= 1
            if matrix[x][y] == target:
                ans += 1
            x += 1

        return ans

if __name__ == '__main__':
    s = Solution()
    print s.searchMatrix([[1, 3, 5, 7],[2, 4, 7, 8],[3, 5, 9, 10]], 3)

File: leetcode-2/search-a-2d-matrix-ii.cc
// Question: https://leetcode.com/problems/search-a-2d-matrix-ii/

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0 || matrix[0].size() == 0) {
            return false;
        }

        int row = matrix.size();
        int col = matrix[0].size();
        int irow = 0, icol = col-1;
        while (irow < row && icol >= 0) {
            if (target == matrix[irow][icol]) {
                return true;
            } else if (target > matrix[irow][icol]) {
                ++irow;
            } else {
                --icol;
            }
        }
        return false;
    }
};
